### PR TITLE
[Issue-3449] Support Ton transfer

### PR DESCRIPTION
**Pull request by bluezdot**

### CODE REVIEW

File: packages/extension-base/src/services/transaction-service/index.ts [hunk 1]
```diff
@@ -1152,6 +1155,88 @@ export default class TransactionService {
...
+        }
+
+        console.log('externalMsgHash', externalMsgHash); // alibaba
+
+        getStatusByExtMsgHash(externalMsgHash).then(([status, hex]) => {
+          console.log('hex', hex); // alibaba
+
+          if (status && hex) {
+            emitter.emit('extrinsicHash', eventData);
+            eventData.extrinsicHash = hex;
```
Comment (S2kael) [inline packages/extension-base/src/services/transaction-service/index.ts:1211]: Move this line up

File: packages/extension-base/src/services/transaction-service/index.ts [hunk 2]
```diff
@@ -1152,6 +1155,88 @@ export default class TransactionService {
...
+            emitter.emit('success', eventData);
+          }
+
+          if (!status && hex) {
+            emitter.emit('extrinsicHash', eventData);
+            eventData.extrinsicHash = hex;
```
Comment (S2kael) [inline packages/extension-base/src/services/transaction-service/index.ts:1217]: Move this line up

File: packages/extension-base/src/services/balance-service/helpers/subscribe/ton/utils.ts [hunk 3]
```diff
@@ -24,18 +25,107 @@ export async function getJettonWalletContract (jettonMasterContract: OpenedContr
...
+      }
+
+      await new Promise((resolve) => setTimeout(resolve, options.delay));
+    }
+  }
+
+  throw lastError;
+}
+
+export async function getStatusByExtMsgHash (extMsgHash: string): Promise<[boolean, string]> {
```
Comment (S2kael) [inline packages/extension-base/src/services/balance-service/helpers/subscribe/ton/utils.ts:104]: Make this function more general and reusable

File: packages/extension-base/src/core/logic-validation/transfer.ts [hunk 4]
```diff
@@ -148,6 +150,8 @@ export async function estimateFeeForTransaction (validationResponse: SWTransacti
     try {
       if (isSubstrateTransaction(transaction)) {
         estimateFee.value = (await transaction.paymentInfo(validationResponse.address)).partialFee.toString();
+      } else if (isTonTransaction(transaction)) {
+        estimateFee.value = transaction.estimateFee;
```
[the remark this answers is not in the document]
Comment (bluezdot) [reply packages/extension-base/src/core/logic-validation/transfer.ts:158]: Might need to update logic estimate fee inside for future actions excluding normal transfer Ton and Jetton

File: packages/extension-base/src/services/balance-service/transfer/ton-transfer.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+      bounce: false // todo: check and update the send bounced logic
+    })
+  ];
+
+  const estimateFee = await estimateTonTxFee(tonApi, messages, walletContract);
+
+  if (transferAll) {
+    const balance = await tonApi.getBalance(tonAddress);
+
+    transferValue = (balance - estimateFee).toString();
```
Comment (nampc1) [inline packages/extension-base/src/services/balance-service/transfer/ton-transfer.ts:67]: I think this should be calculated beforehand

Comment (nampc1) [reply packages/extension-base/src/services/balance-service/transfer/ton-transfer.ts:67]: Max transfer value should be calculated before this step

File: packages/extension-base/src/services/balance-service/transfer/ton-transfer.ts [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    messages,
+    estimateFee: estimateFee.toString(),
+    seqno
+  } as TonTransactionConfig;
+
+  return [transactionObject, transactionObject.value];
+}
+
+async function createJettonTransaction ({ from, networkKey, to, tokenInfo, tonApi, transferAll, value }: TonTransactionConfigProps): Promise<[TonTransactionConfig | null, string]> {
+  const keyPair = keyring.getPair(from);
```
Comment (nampc1) [inline packages/extension-base/src/services/balance-service/transfer/ton-transfer.ts:84]: I don't think we should access keyring directly

File: packages/extension-base/src/services/transaction-service/index.ts [hunk 7]
```diff
@@ -236,7 +243,7 @@ export default class TransactionService {
 
   private async sendTransaction (transaction: SWTransaction): Promise<TransactionEmitter> {
     // Send Transaction
-    const emitter = await (transaction.chainType === 'substrate' ? this.signAndSendSubstrateTransaction(transaction) : this.signAndSendEvmTransaction(transaction));
+    const emitter = await (transaction.chainType === 'substrate' ? this.signAndSendSubstrateTransaction(transaction) : transaction.chainType === 'evm' ? this.signAndSendEvmTransaction(transaction) : this.signAndSendTonTransaction(transaction));
```
Comment (nampc1) [inline packages/extension-base/src/services/transaction-service/index.ts:246]: should refactor for readability

File: packages/extension-base/src/services/balance-service/transfer/ton-transfer.ts (hunk 6 above)
Comment (bluezdot) [reply packages/extension-base/src/services/balance-service/transfer/ton-transfer.ts:87]: The pubKey have to be accessed to get create `wallet contract` - need for estimating tx fee.

File: packages/extension-base/src/services/balance-service/transfer/ton-transfer.ts (hunk 5 above)
Comment (bluezdot) [reply packages/extension-base/src/services/balance-service/transfer/ton-transfer.ts]: I've implemented the max transfer value before. But it's required the sent amount to exactly estimate the txFee. The fee increases a bit with logarithm of sent amount.

Comment (bluezdot) [reply packages/extension-base/src/services/balance-service/transfer/ton-transfer.ts]: -> Solution: Use FreeBalance as the sending amount to estimate fee

**Review by nampc1 — APPROVED** [review]

**Pull request merged**